graphsync/graphsync.md

```diff
@@ -0,0 +1,151 @@
+
+## Interfaces
+
+This is a listing of the data structures and process interfaces involved in the graphsync protocol. For simplicity, we use Go type notation, though of course graphsync is language agnostic.
+
+```go
+type Graphsync interface {
+	Request(req	Request) (Response, error)
+}
+
+type Request struct {
+    Selector Selector
+    Priority Priority      // optional
+    Expires  time.Duration // optional
+}
+
+type GraphSyncNet interface {
+    SendMessage(m Message)
+    RecvMessage(m Message)
+}
+```
+
+
+## Network Messages
+
+```protobuf
+message GraphsyncMessage {
+
+  message Request {
+    int32 id = 1;       // unique id set on the requester side
+    bytes selector = 2; // ipld selector to retrieve
+    bytes extra = 3;    // aux information. useful for other protocols
+    int32 priority = 4;	// the priority (normalized). default to 1
+    bool  cancel = 5;   // whether this cancels a request
+  }
+
+  message RequestList {
+    repeated Request reqs = 1; // a list of requests
+    bool full = 2;			   // whether this is the full RequestList. default to false, for incremental requests.
+  }
+
+  message Response {
+    int32 id = 1;     // the request id
+    int32 status = 2; // a status code.
+    bytes extra = 3;
+  }
+
+  message Block {
+  	bytes prefix = 1; // CID prefix (cid version, multicodec and multihash prefix (type + length)
+  	bytes data = 2;
+  }
+
+  // the actual data included in this message
+  RequestList reqlist = 1;
+  repeated Response reslist = 2;
+  repeated Block data = 2;
+}
+```
+
+### Response Status Codes
+
+```
+# info - partial
+10   Request Acknowledged. Working on it.
+11   Additional Peers. PeerIDs in extra.
+12   Not enough vespene gas ($)
+13   Other Protocol - info in extra.
+
+# success - terminal
+20   Request Completed, full content.
+21   Request Completed, partial content.
+
+# error - terminal
+30   Request Rejected. NOT working on it.
+31   Request failed, busy, try again later (getting dosed. backoff in extra).
+32   Request failed, for unknown reason. Extra may have more info.
+33   Request failed, for legal reasons.
+34   Request failed, content not found.
+```
+
+## Example Use Cases
+
+### Syncing a Blockchain
+
+Requests we would like to make for this:
+
+- Give me `<hash>/Parent`, `<hash>/Parent/Parent` and so on, up to a depth of `N`.
+- Give me nodes that exist in `<hash1>` but not `<hash2>`
+  - In addition to this, the ability to say "Give me some range of (the above query) is very important". For example: "Give me the second 1/3 of the nodes that are children of `<hash1>` but not `<hash2>`"
+
+### Downloading Package Dependencies
+
+- Give me everything within `<hash>/foo/v1.0.0`
+
+### Loading content from deep within a giant dataset
+
+- Give me the nodes for the path `<hash>/a/b/c/d/e/f/g`
+
+### Loading a large video optimizing for playback and seek
```

Comment:
The problem with these approaches that say "give me <indefinite number> of blocks" is that they don't parallelize well.

The client making a request of indefinite size, by definition, doesn't know the size and shape of the request. At each layer of depth we lose the ability to grab that content from other peers. Depending on the shape of the graph and the network conditions of the different peers this approach is rarely the fastest and for many edge cases ends up being incredibly punitive. If we instead just focus on requesting more blocks concurrently once we have the knowledge to do so we actually *gain* concurrency vectors at each depth of the tree and can optimize our requests across many peers.

The inability to request a large number blocks *serially* from a single peer should not be a source of performance issues right now. Currently, we think the largest bottlenecks are:

* inability to grab a deep node without a roundtrip for every branch.
* lack of parallelization when grabbing blocks from a peer (which, on the JS side at least, is apparently just a bug).
* lack of storage knowledge for each CID (basically, we need to note when we have the full graph for a CID so that when the client is syncing an updated graph it knows if it needs to continue traversing down a particular path).

The rough outline of a fix for this was:

* Create a new RPC call that asks for the merkle proof of a given PATH.
* Request large numbers of blocks in parallel from known peers (I think someone said this isn't actually a problem on the Go side).
* Create a layer on top of BlockStore that has a boolean index for whether or not the local repo has the entire graph referenced by that CID.

That's enough to drastically improve over what we have today and avoids a lot of very complex logic when trying to spread out these indefinite requests over multiple peers.